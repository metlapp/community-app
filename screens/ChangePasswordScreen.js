import React, { useEffect, useState } from "react";
import * as Yup from "yup";
import { SafeAreaView, StyleSheet, Text } from "react-native";

import AppFormField from "../components/AppFormField";
import Form from "../components/Form";
import SubmitButton from "../components/SubmitButton";
import ErrorMessage from "../components/ErrorMessage";
import SuccessMessage from "../components/SuccessMessage";
import runCrypto from "../auth/crypto-hashing";

const validationSchema = Yup.object().shape({
  currentPassword: Yup.string().required().min(8).label("Current Password"),
  newPassword: Yup.string().required().min(8).label("New Password"),
  confirmNewPassword: Yup.string().required().min(8).label("Confirm Password"),
});

export default function ChangePasswordScreen() {
  // Temporary user data
  const [user, setUser] = useState({
    id: 4,
    name: "john",
    password:
      "5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8",
  });
  const [errorVisible, setErrorVisible] = useState(false);
  const [error, setError] = useState(false);
  const [successVisible, setSuccessVisible] = useState(false);

  // Reset errors upon page render
  useEffect(() => {
    setSuccessVisible(false);
    setErrorVisible(false);
  }, []);

  // reset form to initial values
  const resetForm = (values) => {
    values["newPassword"] = "";
    values["currentPassword"] = "";
    values["confirmNewPassword"] = "";
  };

  const handleSubmit = async (values) => {
    setErrorVisible(false);
    // Show error if new passwords do not match
    if (values["newPassword"] !== values["confirmNewPassword"]) {
      console.log("passes didnt match");
      setError("Please make sure new passwords match!");
      setErrorVisible(true);
      return;
    }
    values["newPassword"] === values["confirmNewPassword"];
    const currentPassword = await runCrypto(values["currentPassword"]);
    // show error if current password doesnt match password from api
    if (user.password !== currentPassword) {
      setError("Incorrect password");
      setErrorVisible(true);
      return;
    }
    // If no error, submit password change
    const newPassword = await runCrypto(values["newPassword"]);
    setUser({ ...user, password: newPassword });
    setErrorVisible(false);
    setSuccessVisible(true);
    resetForm(values);
    setTimeout(() => {
      setSuccessVisible(false);
    }, 5000);
  };

  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.text}>Change your password</Text>
      <Form
        className="form"
        initialValues={{
          currentPassword: "",
          newPassword: "",
          confirmNewPassword: "",
        }}
        onSubmit={handleSubmit}
        style={styles.form}
        testID="form"
        validationSchema={validationSchema}
      >
        <AppFormField
          id="currentPassword"
          autoFocus
          onFocus={() => setErrorVisible(false)}
          secureTextEntry
          placeholder="Current Password"
          name="currentPassword"
          style={styles.TextInput}
        />
        <AppFormField
          id="newPassword"
          secureTextEntry
          placeholder="New Password"
          name="newPassword"
          style={styles.TextInput}
        />
        <AppFormField
          id="confirmPassword"
          secureTextEntry
          placeholder="Confirm Password"
          name="confirmNewPassword"
          style={styles.TextInput}
        />
        <ErrorMessage
          error={error}
          visible={errorVisible}
          className="errorMessage"
        />
        <SubmitButton className="submit" title="Save">
          Save
        </SubmitButton>
        <SuccessMessage
          success="Password changed successfully"
          visible={successVisible}
        />
      </Form>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
  },
  form: {
    width: "70%",
  },
  text: {
    fontSize: 24,
    fontWeight: "600",
    marginBottom: 30,
    marginTop: 5,
  },
  TextInput: {
    lineHeight: 25,
    fontSize: 20,
    paddingLeft: 10,
  },
});
