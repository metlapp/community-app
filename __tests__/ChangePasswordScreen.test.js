import "react-native";
import React from "react";
import { mount, shallow } from "enzyme";
import ChangePasswordScreen from "../screens/ChangePasswordScreen";

// Note: this is just for use with Jest snapshot testing
// and comes packaged with react-native init project.
// You do not need this if using Enzyme 'toMatchSnapshot' etc.

import renderer from "react-test-renderer";

describe("Testing ChangePasswordScreen", () => {
  it("renders correctly, test using Jest", () => {
    renderer.create(<ChangePasswordScreen />);
  });

  it("matches the snapshot", () => {
    const tree = renderer.create(<ChangePasswordScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("text on change password screen", () => {
    const wrapper = shallow(<ChangePasswordScreen />);
    const text = wrapper.find("Text").render().text();
    expect(text).toEqual("Change your password");
  });

  // it("should display error when required fields are missed", () => {
  //   const wrapper = mount(<ChangePasswordScreen />);

  //   wrapper.find("Form").simulate("submit");

  //   wrapper.debug();
  // });

  it("should say the correct value", () => {
    const wrapper = mount(<ChangePasswordScreen />);

    console.log(wrapper.debug());

    const currentPassword = wrapper.find("TextInput").find("#currentPassword");
    const newPassword = wrapper.find("TextInput").find("#newPassword");
    const confirmPassword = wrapper.find("TextInput").find("#confirmPassword");

    currentPassword.props().value = "Password";
    newPassword.props().value = "Peyton12";
    confirmPassword.props().value = "Peyton12";

    expect(currentPassword.props().value).toEqual("Password");
    expect(newPassword.props().value).toEqual("Peyton12");
    expect(confirmPassword.props().value).toEqual("Peyton12");
  });

  it('should say "Please make sure new passwords match!" when the new passwords dont match', () => {
    const wrapper = mount(<ChangePasswordScreen />);

    const currentPassword = wrapper.find("TextInput").find("#currentPassword");
    const newPassword = wrapper.find("TextInput").find("#newPassword");
    const confirmPassword = wrapper.find("TextInput").find("#confirmPassword");

    const errorMessage = wrapper.find(".errorMessage");

    currentPassword.props().value = "password";
    newPassword.props().value = "Peyton22";
    confirmPassword.props().value = "Peyton12";

    expect(errorMessage.props("error")).toEqual(
      "Please make sure new passwords match!"
    );
  });

  // it('should say "Please make sure new passwords match!" when the new passwords dont match', () => {
  //   const wrapper = shallow(<ChangePasswordScreen />);

  //   const currentPassword = wrapper
  //     .find("#confirmPassword")
  //     .simulate("change", {
  //       value: "password",
  //     });
  //   const newPassword = wrapper.find("#confirmPassword").simulate("change", {
  //     value: "Peyton13",
  //   });
  //   const confirmPassword = wrapper
  //     .find("#confirmPassword")
  //     .simulate("change", {
  //       value: "Peyton12",
  //     });
  //   wrapper.find("Form").simulate("submit");

  //   setTimeout(() => {
  //     const text = wrapper.find("Text").render().text();
  //     expect(text).toEqual("Please make sure new passwords match!");
  //   }, 1000);
  // });

  // it('should say "Password changed successfully" when the password is changed successfully', () => {
  //   const wrapper = shallow(<ChangePasswordScreen />);

  //   const currentPassword = wrapper
  //     .find("#confirmPassword")
  //     .simulate("change", {
  //       value: "password",
  //     });
  //   const newPassword = wrapper.find("#confirmPassword").simulate("change", {
  //     value: "Peyton12",
  //   });
  //   const confirmPassword = wrapper
  //     .find("#confirmPassword")
  //     .simulate("change", {
  //       value: "Peyton12",
  //     });
  //   wrapper.find("Form").simulate("submit");

  //   setTimeout(() => {
  //     const text = wrapper.find("Text").render().text();
  //     expect(text).toEqual("Password changed successfully");
  //   }, 1000);
  // });
});
