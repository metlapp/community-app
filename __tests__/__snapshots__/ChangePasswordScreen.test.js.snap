// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ChangePasswordScreen matches the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "fontWeight": "600",
        "marginBottom": 30,
        "marginTop": 5,
      }
    }
  >
    Change your password
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#f2f2f2",
          "borderRadius": 25,
          "height": 55,
          "justifyContent": "center",
          "marginVertical": 15,
        },
        Object {
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoFocus={true}
      id="currentPassword"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Current Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "fontSize": 20,
          "lineHeight": 25,
          "paddingLeft": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#f2f2f2",
          "borderRadius": 25,
          "height": 55,
          "justifyContent": "center",
          "marginVertical": 15,
        },
        Object {
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      id="newPassword"
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="New Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "fontSize": 20,
          "lineHeight": 25,
          "paddingLeft": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#f2f2f2",
          "borderRadius": 25,
          "height": 55,
          "justifyContent": "center",
          "marginVertical": 15,
        },
        Object {
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      id="confirmPassword"
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Confirm Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "fontSize": 20,
          "lineHeight": 25,
          "paddingLeft": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "blue",
        "borderRadius": 25,
        "justifyContent": "center",
        "marginVertical": 5,
        "opacity": 1,
        "padding": 15,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "fontWeight": "bold",
          "textTransform": "uppercase",
        }
      }
    >
      Save
    </Text>
  </View>
</RCTSafeAreaView>
`;
